File: src/test/java/no/nav/syfo/service/JournalpostServiceTest.java
package no.nav.syfo.service;

import no.nav.syfo.consumer.ws.BehandleInngaaendeJournalConsumer;
import no.nav.syfo.consumer.ws.BehandlendeEnhetConsumer;
import no.nav.syfo.consumer.ws.InngaaendeJournalConsumer;
import no.nav.syfo.consumer.ws.JournalConsumer;
import no.nav.syfo.domain.InngaaendeJournal;
import no.nav.syfo.domain.Inntektsmelding;
import no.nav.syfo.util.Metrikk;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Optional;

import static no.nav.syfo.domain.InngaaendeJournal.MIDLERTIDIG;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class JournalpostServiceTest {

    @Mock
    private InngaaendeJournalConsumer inngaaendeJournalConsumer;
    @Mock
    private BehandlendeEnhetConsumer behandlendeEnhetConsumer;
    @Mock
    private JournalConsumer journalConsumer;
    @Mock
    private BehandleInngaaendeJournalConsumer behandleInngaaendeJournalConsumer;
    @Mock
    private Metrikk metrikk;

    @InjectMocks
    private JournalpostService journalpostService;

    @Test
    public void ferdigstillJournalpost() {
        InngaaendeJournal journal = InngaaendeJournal.builder().dokumentId("dokumentId").status("MIDLERTIDIG").build();
        when(inngaaendeJournalConsumer.hentDokumentId("journalpostId")).thenReturn(journal);

        journalpostService.ferdigstillJournalpost(
                "saksId",
                Inntektsmelding.builder()
                        .fnr("fnr")
                        .arbeidsgiverOrgnummer(Optional.of("orgnummer"))
                        .journalpostId("journalpostId")
                        .arbeidsforholdId(null)
                        .arsakTilInnsending("Ny")
                        .status(MIDLERTIDIG)
                        .build());

        verify(behandlendeEnhetConsumer).hentBehandlendeEnhet("fnr");
        verify(inngaaendeJournalConsumer).hentDokumentId("journalpostId");
        verify(behandleInngaaendeJournalConsumer).oppdaterJournalpost(any());
        verify(behandleInngaaendeJournalConsumer).ferdigstillJournalpost(any());
    }

    @Test
    public void hentInntektsmelding() {
        InngaaendeJournal journal = InngaaendeJournal.builder().dokumentId("dokumentId").status("MIDLERTIDIG").build();
        when(inngaaendeJournalConsumer.hentDokumentId("journalpostId")).thenReturn(journal);
        when(journalConsumer.hentInntektsmelding("journalpostId", journal))
                .thenReturn(Inntektsmelding
                        .builder()
                        .arbeidsgiverOrgnummer(Optional.of("orgnummer"))
                        .orgnummerPrivatperson(Optional.empty())
                        .fnr("fnr")
                        .journalpostId("journalpostId")
                        .status("MIDLERTIDIG")
                        .build());

        Inntektsmelding inntektsmelding = journalpostService.hentInntektsmelding("journalpostId");

        assertThat(inntektsmelding.getFnr()).isEqualTo("fnr");
        assertThat(inntektsmelding.getOrgnummerPrivatperson()).isEqualTo(Optional.empty());
    }
}

File: src/test/kotlin/no/nav/syfo/service/JournalpostServiceTest.kt
package no.nav.syfo.service

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import no.nav.syfo.domain.InngaaendeJournal
import no.nav.syfo.domain.JournalStatus
import no.nav.syfo.domain.inntektsmelding.Inntektsmelding
import no.nav.syfo.util.Metrikk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.LocalDateTime

class JournalpostServiceTest {
    private var inngaaendeJournalConsumer = mockk<InngaaendeJournalConsumer>(relaxed = true)
    private var behandlendeEnhetConsumer = mockk<BehandlendeEnhetConsumer>(relaxed = true)
    private var journalConsumer = mockk<JournalConsumer>(relaxed = true)
    private var behandleInngaaendeJournalConsumer = mockk<BehandleInngaaendeJournalConsumer>(relaxed = true)
    private val metrikk = mockk<Metrikk>(relaxed = true)

    private val journalpostService = JournalpostService(
        inngaaendeJournalConsumer,
        behandleInngaaendeJournalConsumer,
        journalConsumer,
        behandlendeEnhetConsumer,
        metrikk
    )

    @Test
    fun ferdigstillJournalpost() {
        val journal = InngaaendeJournal(dokumentId = "dokumentId", status = JournalStatus.MOTTATT, mottattDato = LocalDateTime.now())
        every { inngaaendeJournalConsumer.hentDokumentId("journalpostId") } returns journal
        every { behandlendeEnhetConsumer.hentBehandlendeEnhet(any(), any()) } returns "enhet"

        journalpostService.ferdigstillJournalpost(
            "saksId",
            Inntektsmelding(
                fnr = "fnr",
                arbeidsgiverOrgnummer = "orgnummer",
                arbeidsforholdId = null,
                journalpostId = "journalpostId",
                arsakTilInnsending = "Ny",
                journalStatus = JournalStatus.MOTTATT,
                arkivRefereranse = "AR123",
                førsteFraværsdag = LocalDate.now(),
                mottattDato = LocalDateTime.now()
            )
        )

        verify { behandlendeEnhetConsumer.hentBehandlendeEnhet("fnr", "") }
        verify { inngaaendeJournalConsumer.hentDokumentId("journalpostId") }
        verify { behandleInngaaendeJournalConsumer.oppdaterJournalpost(any()) }
        verify { behandleInngaaendeJournalConsumer.ferdigstillJournalpost(any()) }
    }

    @Test
    fun hentInntektsmelding() {
        val journal = InngaaendeJournal(dokumentId = "dokumentId", status = JournalStatus.MOTTATT, mottattDato = LocalDateTime.now())
        every { inngaaendeJournalConsumer.hentDokumentId("journalpostId") } returns journal
        every { journalConsumer.hentInntektsmelding("journalpostId", "AR-1234") } returns
            Inntektsmelding(
                fnr = "fnr",
                arbeidsgiverOrgnummer = "orgnummer",
                arbeidsgiverPrivatFnr = null,
                journalpostId = "journalpostId",
                arsakTilInnsending = "",
                journalStatus = JournalStatus.MOTTATT,
                arkivRefereranse = "AR123",
                førsteFraværsdag = LocalDate.now(),
                mottattDato = LocalDateTime.now()
            )


        val (_, fnr, _, arbeidsgiverPrivat) = journalpostService.hentInntektsmelding("journalpostId", "AR-1234")

        assertThat(fnr).isEqualTo("fnr")
        assertThat(arbeidsgiverPrivat).isNull()
    }
}
